### DISCUSSION TITLE
Using v-model for q-select with computed property

### DISCUSSION THREAD

**Discussion by Gabe-S5** [quasarframework/quasar#10987] [General - Components / Directives / etc] [answered]
I currently have a q-select with a v-model of a computed property that returns a prop object. This prop comes from an asynchronous api response for a parent component that has an initial value for the q-select. It seems that the computed property is updated once the api response arrives with the data and presents it into the q-select. 

```
<template>
    <div v-for="node in segment" :key="node.id">
          <q-select
            clearable
            v-model="tickedModel[node.label]"
            :options="getOptions(node.children)"
            :label="node.label"
            color="secondary"
            new-value-mode="toggle"
            @update:model-value="updateTicked(tickedModel)"
          />
    </div>
<template/>

<script>
  props: ["tickState"],
  computed: {
    tickedModel() {
      return this.tickState;
    },
<script/>
```
The code works as intended and the q-select has its initial state and I am also able to change the selected option. 

Question: Why does this work?

From my understanding, the computed property should be read only and that a computed property only run when any values within are changed. With this in mind my tickedModel would always return this.tickState and I should not be able to modify it. Therefore I should not be able to select a new option since the model is not allowed to be mutated. Clarification of my understanding of using computed properties would be appreciated.

**Comment by pdanpdan** [comment]
Show a sample of `tickState`

In fact it doesn't matter - I suppose it's an array, and you pass one item in the array to `v-model` - you do not change the array as a whole but only an item inside it, so vue will not error that you are trying to change it.

But this is `a very bad thing` :tm: because you'll end up not knowing who did what and where.

**Reply by Gabe-S5** [reply]
```
{Marital Status: 'In a relationship', Education: "Bachelor's Degree", Home-ownership Status: '', Parental Status: ''}
```
This is a sample initial tickState. 
![image](https://user-images.githubusercontent.com/77077817/137370541-6a914655-05f1-411e-a023-6eb8bf934733.png)

**Reply by pdanpdan** [reply]
Ok, it's an object not an array, but the same goes.

If you do
```javascript
computed: {
    tickedModel() {
      return Object.freeze(this.tickState);
    },
},
```
you'll see the error.

**Reply by MichaelMarkieta** [reply]
Advice on the pattern or approach we should take? Here's the high level architecture:

1. async API call is made by a parent component
2. child component received data from API call as a prop
3. q-select needs the result of the API response from the prop, but props should not be used as the v-model for q-select

**Comment by pdanpdan** [comment, marked as answer]
```vue
<template>
  <div v-for="node in segment" :key="node.id">
    <q-select
      clearable
      :model-value="tickState[node.label]"
      :options="getOptions(node.children)"
      :label="node.label"
      color="secondary"
      new-value-mode="toggle"
      @update:model-value="val => { setModel(node.label, val) }"
    />
  </div>
<template/>

<script>
export default {
  props: ['tickState', 'segment'],

  emits: ['update:tickState'],

  methods: {
    setModel (key, value) {
      this.$emit('update:tickState', {
        ...this.tickState,
        [key]: value
      })
    },

    getOptions (list) {
      return list
    }
  }
}
<script/>
```

and use it from outside:
```vue
<template>
  <my-component v-model:tick-state="myList" />
<template/>
```
